<a href="https://colab.research.google.com/github/isaacadebayo/Coding-Projects/blob/main/MSDS570_Final_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


**DATA HEADING**
*   Prevalence of Undernourishment (% of the population) - Prevalence of undernourishment is the percentage of the population whose habitual food consumption is insufficient to provide the dietary energy levels that are required to maintain a normally active and healthy life.

*   Carbon dioxide emissions (kiloton) - Carbon dioxide emissions are those stemming from the burning of fossil fuels and the manufacture of cement. They include carbon dioxide produced during the consumption of solid, liquid, and gas fuels and gas flaring

*   Health Expenditure (% of GDP) - Level of current health expenditure expressed as a percentage of GDP. Estimates of current health expenditures include healthcare goods and services consumed during each year. This indicator does not include capital health expenditures such as buildings, machinery, IT, and stocks of vaccines for emergencies or outbreaks.

*   Education Expenditure (% of GDP) - General government expenditure on education (current, capital, and transfers) is expressed as a percentage of GDP. It includes expenditures funded by transfers from international sources to the government. General government usually refers to local, regional, and central governments.

*   Unemployment (% total labor force) - Unemployment refers to the % share of the labor force that is without work but available for and seeking employment
Corruption (CPIA rating) - Transparency, accountability, and corruption in the public sector assets the extent to which the executive can be held accountable for its use of funds and for the results of its actions by the electorate and by the legislature and judiciary, and the extent to which public employees within the executive are required to account for administrative decisions, use of resources, and results obtained.
*   Sanitation - People using safely managed sanitation services (% of the population): The percentage of people using improved sanitation facilities that are not shared with other households and where excreta are safely disposed of in situ or transported and treated offsite. Improved sanitation facilities include flush/pour flush to piped sewer systems, septic tanks, or pit latrines: ventilated improved pit latrines, compositing toilets, or pit latrines with slabs. WHO/UNICEF Joint Monitoring Programme (JMP) for Water Supply, Sanitation and Hygiene (washdata.org).

*   Disability-Adjusted Life Years (DALYs) due to Injuries - One DALY represents the loss of the equivalent of one year of full health. DALYs for an injury or health condition is the sum of the years of life lost due to premature mortality (YLLs) and the years lived with a disability (YLDs) due to prevalent cases of the disease in a population

*   Disability-Adjusted Life Years (DALYs) due to Communicable diseases - One DALY represents the loss of the equivalent of one year of full health. DALYs for a communicable disease or health condition is the sum of the years of life lost due to premature mortality (YLLs) and the years lived with a disability (YLDs) due to prevalent cases of the disease in a population

*   Disability-Adjusted Life Years (DALYs) due to Non-Communicable diseases - One DALY represents the loss of the equivalent of one year of full health. DALYs for a non-communicable disease or health condition is the sum of the years of life lost due to premature mortality (YLLs) and the years lived with a disability (YLDs) due to prevalent cases of the disease in a populationt**   











In [69]:
import pandas as pd
import numpy as np

In [70]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [71]:
# reading data
data = pd.read_csv('/content/drive/MyDrive/life_expectancy.csv')
data.head()

,Country Name,Country Code,Region,IncomeGroup,Year,Life Expectancy World Bank,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Corruption,Sanitation,Injuries,Communicable,NonCommunicable
0,Afghanistan,AFG,South Asia,Low income,2001,56.308,47.8,730.0,NaN,NaN,10.809000,NaN,NaN,2179727.10,9689193.70,5795426.38
1,Angola,AGO,Sub-Saharan Africa,Lower middle income,2001,47.059,67.5,15960.0,4.483516,NaN,4.004000,NaN,NaN,1392080.71,11190210.53,2663516.34
2,Albania,ALB,Europe & Central Asia,Upper middle income,2001,74.288,4.9,3230.0,7.139524,3.4587,18.575001,NaN,40.520895,117081.67,140894.78,532324.75
3,Andorra,AND,Europe & Central Asia,High income,2001,NaN,NaN,520.0,5.865939,NaN,NaN,NaN,21.788660,1697.99,695.56,13636.64
4,United Arab Emirates,ARE,Middle East & North Africa,High income,2001,74.544,2.8,97200.0,2.484370,NaN,2.493000,NaN,NaN,144678.14,65271.91,481740.70


In [72]:
data.describe()

,Year,Life Expectancy World Bank,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Corruption,Sanitation,Injuries,Communicable,NonCommunicable
count,3306.000000,3118.000000,2622.000000,3.154000e+03,3126.000000,2216.000000,3002.000000,975.000000,2059.000000,3.306000e+03,3.306000e+03,3.306000e+03
mean,2010.000000,69.748362,10.663654,1.574924e+05,6.364059,4.589014,7.890760,2.860513,52.738785,1.318219e+06,4.686289e+06,7.392488e+06
std,5.478054,9.408154,11.285897,7.726415e+05,2.842844,2.119165,6.270832,0.621343,30.126762,5.214068e+06,1.843727e+07,2.932688e+07
min,2001.000000,40.369000,2.500000,1.000000e+01,1.263576,0.850320,0.100000,1.000000,2.377647,4.304900e+02,3.301600e+02,2.481820e+03
25%,2005.000000,63.642000,2.500000,2.002500e+03,4.205443,3.136118,3.733000,2.500000,24.746007,6.245688e+04,5.776475e+04,3.184758e+05
50%,2010.000000,72.168500,6.200000,1.020500e+04,5.892352,4.371465,5.920000,3.000000,49.317481,2.456910e+05,3.147693e+05,1.350146e+06
75%,2015.000000,76.809000,14.775000,5.877250e+04,8.119166,5.519825,10.097500,3.250000,80.278847,8.465591e+05,2.831636e+06,3.918468e+06
max,2019.000000,84.356341,70.900000,1.070722e+07,24.230680,23.270000,37.250000,4.500000,100.000004,5.563676e+07,2.685646e+08,3.246378e+08


**DATA INSPECTION**

In [73]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 16 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country Name                    3306 non-null   object 
 1   Country Code                    3306 non-null   object 
 2   Region                          3306 non-null   object 
 3   IncomeGroup                     3306 non-null   object 
 4   Year                            3306 non-null   int64  
 5   Life Expectancy World Bank      3118 non-null   float64
 6   Prevelance of Undernourishment  2622 non-null   float64
 7   CO2                             3154 non-null   float64
 8   Health Expenditure %            3126 non-null   float64
 9   Education Expenditure %         2216 non-null   float64
 10  Unemployment                    3002 non-null   float64
 11  Corruption                      975 non-null    float64
 12  Sanitation                      20

3 data types: Object, integer and float (useful for knowing thee type of visualization plot)

In [74]:
print(data.shape)

(3306, 16)


In [75]:
data.columns

Index(['Country Name', 'Country Code', 'Region', 'IncomeGroup', 'Year',
       'Life Expectancy World Bank', 'Prevelance of Undernourishment', 'CO2',
       'Health Expenditure %', 'Education Expenditure %', 'Unemployment',
       'Corruption', 'Sanitation', 'Injuries', 'Communicable',
       'NonCommunicable'],
      dtype='object')

Data consists of 3306 rows and 16 Columns

Shows which column / row has a missing variable

In [76]:
data.isna().any()

,0
Country Name,False
Country Code,False
Region,False
IncomeGroup,False
Year,False
Life Expectancy World Bank,True
Prevelance of Undernourishment,True
CO2,True
Health Expenditure %,True
Education Expenditure %,True


In [77]:
data.isna().sum()

,0
Country Name,0
Country Code,0
Region,0
IncomeGroup,0
Year,0
Life Expectancy World Bank,188
Prevelance of Undernourishment,684
CO2,152
Health Expenditure %,180
Education Expenditure %,1090


Tells you how many missing variable are there in each column for instance the corruption column has 2331 missing variable of the 3306 row

In [78]:
catg_data = data.select_dtypes(include = 'object')
catg_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Country Name  3306 non-null   object
 1   Country Code  3306 non-null   object
 2   Region        3306 non-null   object
 3   IncomeGroup   3306 non-null   object
dtypes: object(4)
memory usage: 103.4+ KB


In [79]:
num_data = data.select_dtypes(exclude = 'object')
num_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 12 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Year                            3306 non-null   int64  
 1   Life Expectancy World Bank      3118 non-null   float64
 2   Prevelance of Undernourishment  2622 non-null   float64
 3   CO2                             3154 non-null   float64
 4   Health Expenditure %            3126 non-null   float64
 5   Education Expenditure %         2216 non-null   float64
 6   Unemployment                    3002 non-null   float64
 7   Corruption                      975 non-null    float64
 8   Sanitation                      2059 non-null   float64
 9   Injuries                        3306 non-null   float64
 10  Communicable                    3306 non-null   float64
 11  NonCommunicable                 3306 non-null   float64
dtypes: float64(11), int64(1)
memory us

Created two new variable for categorical and numerical and split them into two

In [80]:
num_data['Communicable'].count()

np.int64(3306)

In [81]:
num_data['Communicable'].mean()

np.float64(4686288.641721114)

In [82]:
catg_data['Country Name'].sort_values()

,Country Name
0,Afghanistan
2610,Afghanistan
1566,Afghanistan
1740,Afghanistan
2262,Afghanistan
...,...
521,Zimbabwe
1391,Zimbabwe
2609,Zimbabwe
3131,Zimbabwe


In [83]:
num_data['Year'].sort_values()

,Year
0,2001
1,2001
50,2001
51,2001
52,2001
...,...
3275,2019
3303,2019
3304,2019
3305,2019


Separate each column (country name) by year. Then arrange by region, split, categorical and numerical data again and take the mean for each numerical data by region.

Finally fill NaN to reduce outlier. The column life expectancy is the target variable.

In [84]:
data_drop_corruption = data.copy()
data_drop_corruption

,Country Name,Country Code,Region,IncomeGroup,Year,Life Expectancy World Bank,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Corruption,Sanitation,Injuries,Communicable,NonCommunicable
0,Afghanistan,AFG,South Asia,Low income,2001,56.308,47.8,730.000000,NaN,NaN,10.809000,NaN,NaN,2179727.10,9689193.70,5795426.38
1,Angola,AGO,Sub-Saharan Africa,Lower middle income,2001,47.059,67.5,15960.000000,4.483516,NaN,4.004000,NaN,NaN,1392080.71,11190210.53,2663516.34
2,Albania,ALB,Europe & Central Asia,Upper middle income,2001,74.288,4.9,3230.000000,7.139524,3.45870,18.575001,NaN,40.520895,117081.67,140894.78,532324.75
3,Andorra,AND,Europe & Central Asia,High income,2001,NaN,NaN,520.000000,5.865939,NaN,NaN,NaN,21.788660,1697.99,695.56,13636.64
4,United Arab Emirates,ARE,Middle East & North Africa,High income,2001,74.544,2.8,97200.000000,2.484370,NaN,2.493000,NaN,NaN,144678.14,65271.91,481740.70
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3301,Vanuatu,VUT,East Asia & Pacific,Lower middle income,2019,70.474,12.4,209.999993,3.360347,1.77788,1.801000,3.0,NaN,12484.18,26032.56,69213.56
3302,Samoa,WSM,East Asia & Pacific,Lower middle income,2019,73.321,4.4,300.000012,6.363094,4.70625,8.406000,4.0,47.698788,6652.84,9095.19,43798.62
3303,South Africa,ZAF,Sub-Saharan Africa,Upper middle income,2019,64.131,6.3,439640.014648,9.109355,5.91771,28.469999,NaN,NaN,3174676.10,13198944.71,10214261.89
3304,Zambia,ZMB,Sub-Saharan Africa,Low income,2019,63.886,NaN,6800.000191,5.312203,4.46518,12.520000,2.5,NaN,510982.75,4837094.00,2649687.82


In [85]:
data_drop = data_drop_corruption.drop(columns = 'Corruption')

**NOTE iloc can slice multiple columns of different data type. To slice column that are in order. Just use iloc and the index number e.g (df.iloc[: 0:3]), to slice column in random order use (iloc [:, [0, 3]), 0 being the first index you wish to slice and 3 bing the order index you ned to slice.**:

NOTE: to drop column (differentiate between dropna (used to df.dropna in a row and df.drop which i used to drop a column (put the colum in quotation))). Since i forgot how to do rudimentary things like markup to do a markupo use command or (esc in window ) and M

In [86]:
data.iloc[:, [2, 5]]

,Region,Life Expectancy World Bank
0,South Asia,56.308
1,Sub-Saharan Africa,47.059
2,Europe & Central Asia,74.288
3,Europe & Central Asia,NaN
4,Middle East & North Africa,74.544
...,...,...
3301,East Asia & Pacific,70.474
3302,East Asia & Pacific,73.321
3303,Sub-Saharan Africa,64.131
3304,Sub-Saharan Africa,63.886


In [87]:
data_drop.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 15 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country Name                    3306 non-null   object 
 1   Country Code                    3306 non-null   object 
 2   Region                          3306 non-null   object 
 3   IncomeGroup                     3306 non-null   object 
 4   Year                            3306 non-null   int64  
 5   Life Expectancy World Bank      3118 non-null   float64
 6   Prevelance of Undernourishment  2622 non-null   float64
 7   CO2                             3154 non-null   float64
 8   Health Expenditure %            3126 non-null   float64
 9   Education Expenditure %         2216 non-null   float64
 10  Unemployment                    3002 non-null   float64
 11  Sanitation                      2059 non-null   float64
 12  Injuries                        33

Value counts is used to add up similar value in the same column
".value_counts()"

# LIFE EXPECTANCY

In [88]:
data_merge_life_expect = data_drop.iloc[:, [2, 4, 5]]
data_merge_life_expect.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Region                      3306 non-null   object 
 1   Year                        3306 non-null   int64  
 2   Life Expectancy World Bank  3118 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [89]:
target_data_nan_life_expect = data_drop.groupby(['Region', 'Year']) ['Life Expectancy World Bank'].transform(lambda x: x.fillna(x.mean()))

In [90]:
target_data_nan_life_expect.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Life Expectancy World Bank
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# PREVELANCE OF UNDERNOURISHMENT

In [91]:
data_merge_prevalence_of_under = data_drop.iloc[:, [2, 4, 6]]
data_merge_prevalence_of_under.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Region                          3306 non-null   object 
 1   Year                            3306 non-null   int64  
 2   Prevelance of Undernourishment  2622 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [92]:
target_data_nan_prevalence_of_under = data_drop.groupby(['Region', 'Year']) ['Prevelance of Undernourishment'].transform(lambda x: x.fillna(x.mean()))
# created a new variable called target_data_nan_prevalence_of_under  now do this for every other column with nan using region and year as constant

In [93]:
target_data_nan_prevalence_of_under.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Prevelance of Undernourishment
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# HEALTH EXPENDITURE

In [94]:
data_merge_health_expenditure = data_drop.iloc[:, [2, 4, 8]]
data_merge_health_expenditure.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Region                3306 non-null   object 
 1   Year                  3306 non-null   int64  
 2   Health Expenditure %  3126 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [95]:
target_data_nan_health_expenditure = data_drop.groupby(['Region', 'Year']) ['Health Expenditure %'].transform(lambda x: x.fillna(x.mean()))

In [96]:
target_data_nan_health_expenditure.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Health Expenditure %
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# EDUCATION EXPENDITURE

In [97]:
data_merge_education_expenditure = data_drop.iloc[:, [2, 4, 9]]
# 1090 nan, contemplating whether filling this nan is even worth it.
data_merge_education_expenditure.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Region                   3306 non-null   object 
 1   Year                     3306 non-null   int64  
 2   Education Expenditure %  2216 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [98]:
target_data_nan_education_expenditure = data_drop.groupby(['Region', 'Year']) ['Education Expenditure %'].transform(lambda x: x.fillna(x.mean()))

In [99]:
# Logic error using sort_index
#NAN still present in 12 rows despite filling it based on nan from previous region. So i will
# attempt to fill it from average mean from previous year based on the region

target_data_nan_education_expenditure.info()
data_drop['Education Expenditure %'].isna().sum()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Education Expenditure %
Non-Null Count  Dtype  
--------------  -----  
3294 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


np.int64(1090)

Use forward and backward fill to fill missing values in the column after filling by mean in the region.

In [100]:
# Re-apply the group-wise mean imputation for 'Education Expenditure %'
data_drop['Education Expenditure %'] = data_drop.groupby(['Region', 'Year'])['Education Expenditure %'].transform(lambda x: x.fillna(x.mean()))

data_drop['Education Expenditure %'].isna().sum()

np.int64(12)

Filling NaN with backward and forward fill from previous region or yeear

In [101]:
# Forward fill ffill()
data_drop['Education Expenditure %'] = data_drop.groupby('Region')['Education Expenditure %'].transform(lambda x: x.sort_values().ffill())

# Backward fill bfill()
data_drop['Education Expenditure %'] = data_drop.groupby('Region')['Education Expenditure %'].transform(lambda x: x.sort_values().bfill())

# Backward fill
#data_drop['Education Expenditure %'].fillna(data_drop['Education Expenditure %'].mean(), inplace=True)

target_data_nan_education_expenditure_updated = data_drop['Education Expenditure %']

In [102]:
target_data_nan_education_expenditure_updated.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Education Expenditure %
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


In [103]:
target_data_nan_education_expenditure.isna().any()

np.True_

In [104]:
target_data_nan_education_expenditure_updated.isna().any()

np.False_

# UNEMPLOYMENT

In [105]:
data_merge_unemployment = data_drop.iloc[:, [2, 4, 10]]
# similar to the previous column, 1247 nan. Thinking if i should fill this nan
data_merge_unemployment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Region        3306 non-null   object 
 1   Year          3306 non-null   int64  
 2   Unemployment  3002 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [106]:
target_data_nan_unemployment = data_drop.groupby(['Region', 'Year']) ['Unemployment'].transform(lambda x: x.fillna(x.mean()))

In [107]:
target_data_nan_unemployment.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Unemployment
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# SANITATION

In [108]:
data_merge_sanitation = data_drop.iloc[:, [2, 4, 11]]
# similar to the previous column, 1247 nan. Thinking if i should fill this nan
data_merge_sanitation.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Region      3306 non-null   object 
 1   Year        3306 non-null   int64  
 2   Sanitation  2059 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 77.6+ KB


In [109]:
target_data_nan_sanitation = data_drop.groupby(['Region', 'Year']) ['Sanitation'].transform(lambda x: x.fillna(x.mean()))

It does seem redudant to use groupby after using iloc but I am trying to divide based on region so I can fill nan based on that specific region while iloc just merge cells

In [110]:
target_data_nan_sanitation.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Sanitation
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# CO2

In [111]:
data_merge_CO2 = data_drop.iloc[:, [2, 4, 7]]
data_merge_CO2.isna().info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   Region  3306 non-null   bool 
 1   Year    3306 non-null   bool 
 2   CO2     3306 non-null   bool 
dtypes: bool(3)
memory usage: 9.8 KB


In [112]:
target_data_nan_CO2 = data_drop.groupby(['Region', 'Year']) ['CO2'].transform(lambda x: x.fillna(x.mean()))
# created a new variable called target_data_nan_CO2 now do this for every other column with nan using region and year as constant

AI ask me to use .reset_index() to convert region and year to three different column. I know if my code is bug free from logic error when I use .info to inspect

In [113]:
target_data_nan_CO2.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: CO2
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


# JOINING ALL COLUMN THAT WAS FILLED WITH NAN BEFORE RUNNING SOME MORE INSPECTION

In [114]:
data_drop['CO2'] = target_data_nan_CO2
data_drop['Education Expenditure %'] = target_data_nan_education_expenditure_updated
data_drop['Health Expenditure %'] = target_data_nan_health_expenditure
data_drop['Prevelance of Undernourishment'] = target_data_nan_prevalence_of_under
data_drop['Unemployment'] = target_data_nan_unemployment
data_drop['Life Expectancy World Bank'] = target_data_nan_life_expect
data_drop['Sanitation'] = target_data_nan_sanitation

In [115]:
data_drop.info()

#12 NAN as expected while filling in the nan info showed me some nan were not filled

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 15 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country Name                    3306 non-null   object 
 1   Country Code                    3306 non-null   object 
 2   Region                          3306 non-null   object 
 3   IncomeGroup                     3306 non-null   object 
 4   Year                            3306 non-null   int64  
 5   Life Expectancy World Bank      3306 non-null   float64
 6   Prevelance of Undernourishment  3306 non-null   float64
 7   CO2                             3306 non-null   float64
 8   Health Expenditure %            3306 non-null   float64
 9   Education Expenditure %         3306 non-null   float64
 10  Unemployment                    3306 non-null   float64
 11  Sanitation                      3306 non-null   float64
 12  Injuries                        33

In [116]:
data_drop.describe()

,Year,Life Expectancy World Bank,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Sanitation,Injuries,Communicable,NonCommunicable
count,3306.000000,3306.000000,3306.000000,3.306000e+03,3306.000000,3306.000000,3306.000000,3306.000000,3.306000e+03,3.306000e+03,3.306000e+03
mean,2010.000000,69.959836,11.055657,1.767375e+05,6.398471,4.587310,7.751613,46.339017,1.318219e+06,4.686289e+06,7.392488e+06
std,5.478054,9.193461,10.622629,7.843258e+05,2.813426,1.776753,6.046512,27.847346,5.214068e+06,1.843727e+07,2.932688e+07
min,2001.000000,40.369000,2.500000,1.000000e+01,1.263576,0.850320,0.100000,2.377647,4.304900e+02,3.301600e+02,2.481820e+03
25%,2005.000000,64.215000,2.600000,2.182500e+03,4.304727,3.597885,3.730250,19.647712,6.245688e+04,5.776475e+04,3.184758e+05
50%,2010.000000,72.383051,7.400000,1.208000e+04,5.987030,4.507864,5.956000,42.149323,2.456910e+05,3.147693e+05,1.350146e+06
75%,2015.000000,76.668067,16.200000,7.136000e+04,8.051912,5.294403,9.873250,69.659404,8.465591e+05,2.831636e+06,3.918468e+06
max,2019.000000,84.356341,70.900000,1.070722e+07,24.230680,23.270000,37.250000,100.000004,5.563676e+07,2.685646e+08,3.246378e+08


# Adding all the filled NAN using lambda and transform for all column

A lambda function is a small, anonymous function. It's often used for simple, one-line operations where you don't need to define a full, named function using def.

.transform() is a special method in pandas that is typically used after a groupby() operation. Its key characteristic is that it returns a Series or DataFrame that has the same index and size as the original DataFrame or Series it was applied to.

Role: While a standard aggregation like .mean() or .sum() after a groupby() returns a single value per group (or one row per group), .transform() applies a function to each group and then 'broadcasts' the result back so that each original row gets a value corresponding to its group.

############################################################

############################################################

*****************************************************************************
# PREDICTIVE MODELLING
*****************************************************************************

In [117]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split

In [118]:
longevity_data_predict = data_drop.drop(columns=['Life Expectancy World Bank'])
y = data_drop['Life Expectancy World Bank']

In [119]:
y.info()

<class 'pandas.core.series.Series'>
RangeIndex: 3306 entries, 0 to 3305
Series name: Life Expectancy World Bank
Non-Null Count  Dtype  
--------------  -----  
3306 non-null   float64
dtypes: float64(1)
memory usage: 26.0 KB


In [120]:
longevity_data_predict.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Data columns (total 14 columns):
 #   Column                          Non-Null Count  Dtype  
---  ------                          --------------  -----  
 0   Country Name                    3306 non-null   object 
 1   Country Code                    3306 non-null   object 
 2   Region                          3306 non-null   object 
 3   IncomeGroup                     3306 non-null   object 
 4   Year                            3306 non-null   int64  
 5   Prevelance of Undernourishment  3306 non-null   float64
 6   CO2                             3306 non-null   float64
 7   Health Expenditure %            3306 non-null   float64
 8   Education Expenditure %         3306 non-null   float64
 9   Unemployment                    3306 non-null   float64
 10  Sanitation                      3306 non-null   float64
 11  Injuries                        3306 non-null   float64
 12  Communicable                    33

In [121]:
longevity_data_predict.isna().any()

,0
Country Name,False
Country Code,False
Region,False
IncomeGroup,False
Year,False
Prevelance of Undernourishment,False
CO2,False
Health Expenditure %,False
Education Expenditure %,False
Unemployment,False


# Splitting updated data_drop into categorical and numerical variables

In [122]:
longevity_data_num = longevity_data_predict.select_dtypes(exclude=object)
longevity_data_cat = longevity_data_predict.select_dtypes(include=object)

In [123]:
longevity_data_num

,Year,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Sanitation,Injuries,Communicable,NonCommunicable
0,2001,47.800000,730.000000,4.321165,3.932013,10.809000,25.078307,2179727.10,9689193.70,5795426.38
1,2001,67.500000,15960.000000,4.483516,3.767954,4.004000,11.166260,1392080.71,11190210.53,2663516.34
2,2001,4.900000,3230.000000,7.139524,3.458700,18.575001,40.520895,117081.67,140894.78,532324.75
3,2001,4.814634,520.000000,5.865939,4.713205,10.404860,21.788660,1697.99,695.56,13636.64
4,2001,2.800000,97200.000000,2.484370,5.593494,2.493000,51.930795,144678.14,65271.91,481740.70
...,...,...,...,...,...,...,...,...,...,...
3301,2019,12.400000,209.999993,3.360347,1.777880,1.801000,59.353110,12484.18,26032.56,69213.56
3302,2019,4.400000,300.000012,6.363094,4.706250,8.406000,47.698788,6652.84,9095.19,43798.62
3303,2019,6.300000,439640.014648,9.109355,5.917710,28.469999,20.148443,3174676.10,13198944.71,10214261.89
3304,2019,20.450000,6800.000191,5.312203,4.465180,12.520000,20.148443,510982.75,4837094.00,2649687.82


In [124]:
longevity_data_num.columns

Index(['Year', 'Prevelance of Undernourishment', 'CO2', 'Health Expenditure %',
       'Education Expenditure %', 'Unemployment', 'Sanitation', 'Injuries',
       'Communicable', 'NonCommunicable'],
      dtype='object')

In [125]:
longevity_data_cat

,Country Name,Country Code,Region,IncomeGroup
0,Afghanistan,AFG,South Asia,Low income
1,Angola,AGO,Sub-Saharan Africa,Lower middle income
2,Albania,ALB,Europe & Central Asia,Upper middle income
3,Andorra,AND,Europe & Central Asia,High income
4,United Arab Emirates,ARE,Middle East & North Africa,High income
...,...,...,...,...
3301,Vanuatu,VUT,East Asia & Pacific,Lower middle income
3302,Samoa,WSM,East Asia & Pacific,Lower middle income
3303,South Africa,ZAF,Sub-Saharan Africa,Upper middle income
3304,Zambia,ZMB,Sub-Saharan Africa,Low income


In [126]:
longevity_data_cat.columns

Index(['Country Name', 'Country Code', 'Region', 'IncomeGroup'], dtype='object')

# SCALING

In [127]:
from sklearn.preprocessing import StandardScaler

Scaling to minimize outliers or feature dominance (greater data input with larger range of data input with minor range)

In [128]:
scaler = StandardScaler()
scaled_data = scaler.fit_transform(longevity_data_num)

longevity_scaled_data = pd.DataFrame(longevity_data_predict, columns=longevity_data_num.columns)
display(longevity_scaled_data.describe().T)
display(longevity_scaled_data.corr())


,count,mean,std,min,25%,50%,75%,max
Year,3306.0,2.010000e+03,5.478054e+00,2001.000000,2005.000000,2.010000e+03,2.015000e+03,2.019000e+03
Prevelance of Undernourishment,3306.0,1.105566e+01,1.062263e+01,2.500000,2.600000,7.400000e+00,1.620000e+01,7.090000e+01
CO2,3306.0,1.767375e+05,7.843258e+05,10.000000,2182.500064,1.208000e+04,7.136000e+04,1.070722e+07
Health Expenditure %,3306.0,6.398471e+00,2.813426e+00,1.263576,4.304727,5.987030e+00,8.051912e+00,2.423068e+01
Education Expenditure %,3306.0,4.587310e+00,1.776753e+00,0.850320,3.597885,4.507864e+00,5.294403e+00,2.327000e+01
Unemployment,3306.0,7.751613e+00,6.046512e+00,0.100000,3.730250,5.956000e+00,9.873250e+00,3.725000e+01
Sanitation,3306.0,4.633902e+01,2.784735e+01,2.377647,19.647712,4.214932e+01,6.965940e+01,1.000000e+02
Injuries,3306.0,1.318219e+06,5.214068e+06,430.490000,62456.877500,2.456910e+05,8.465591e+05,5.563676e+07
Communicable,3306.0,4.686289e+06,1.843727e+07,330.160000,57764.745000,3.147693e+05,2.831636e+06,2.685646e+08
NonCommunicable,3306.0,7.392488e+06,2.932688e+07,2481.820000,318475.837500,1.350146e+06,3.918468e+06,3.246378e+08


,Year,Prevelance of Undernourishment,CO2,Health Expenditure %,Education Expenditure %,Unemployment,Sanitation,Injuries,Communicable,NonCommunicable
Year,1.000000,-0.158658,0.022931,0.095742,0.035484,-0.052354,0.122392,-0.005951,-0.037286,0.018347
Prevelance of Undernourishment,-0.158658,1.000000,-0.134534,-0.225205,-0.210425,-0.051497,-0.560493,0.007319,0.141010,-0.043661
CO2,0.022931,-0.134534,1.000000,0.150118,-0.047197,-0.059839,0.150494,0.617354,0.192460,0.781801
Health Expenditure %,0.095742,-0.225205,0.150118,1.000000,0.323792,0.103084,0.257368,-0.088396,-0.175460,-0.047899
Education Expenditure %,0.035484,-0.210425,-0.047197,0.323792,1.000000,0.078521,0.182490,-0.079957,-0.119646,-0.078061
Unemployment,-0.052354,-0.051497,-0.059839,0.103084,0.078521,1.000000,-0.103079,-0.067374,-0.090563,-0.078735
Sanitation,0.122392,-0.560493,0.150494,0.257368,0.182490,-0.103079,1.000000,-0.063428,-0.188065,-0.003699
Injuries,-0.005951,0.007319,0.617354,-0.088396,-0.079957,-0.067374,-0.063428,1.000000,0.773388,0.952623
Communicable,-0.037286,0.141010,0.192460,-0.175460,-0.119646,-0.090563,-0.188065,0.773388,1.000000,0.619248
NonCommunicable,0.018347,-0.043661,0.781801,-0.047899,-0.078061,-0.078735,-0.003699,0.952623,0.619248,1.000000


Strength of Relationship:

Values closer to +1 (e.g., 0.8, 0.9, 0.95) indicate a very strong positive linear relationship. This means that as the values of one column increase, the values of the other column tend to increase very predictably and consistently.
Values closer to -1 (e.g., -0.8, -0.9, -0.95) indicate a very strong negative linear relationship.


This means that as the values of one column increase, the values of the other column tend to decrease very predictably and consistently.

Values closer to 0 (e.g., -0.1, 0.05, 0.15) indicate a weak or no linear relationship. The changes in one column don't seem to have a consistent linear pattern with the changes in the othe

# Creating Dummies for the categorical variable

One hot encoding categorical data  to plug into machine learning algorithm
that uses only binary data

In [129]:
longevity_data_catg_dum = pd.DataFrame()
for c in longevity_data_cat.columns:
    dummy = pd.get_dummies(longevity_data_cat[c], prefix=c, dtype=int)
    dummy_sum = dummy.sum(axis=0)
    col_to_drop = dummy_sum.idxmax()
    dummy = dummy.drop(columns=[col_to_drop])
    longevity_data_catg_dum = pd.concat([longevity_data_catg_dum, dummy], axis=1)

longevity_data_catg_dum

,Country Name_Albania,Country Name_Algeria,Country Name_American Samoa,Country Name_Andorra,Country Name_Angola,Country Name_Antigua and Barbuda,Country Name_Argentina,Country Name_Armenia,Country Name_Australia,Country Name_Austria,...,Country Code_ZWE,Region_East Asia & Pacific,Region_Latin America & Caribbean,Region_Middle East & North Africa,Region_North America,Region_South Asia,Region_Sub-Saharan Africa,IncomeGroup_Low income,IncomeGroup_Lower middle income,IncomeGroup_Upper middle income
0,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,1,0,1,0,0
1,0,0,0,0,1,0,0,0,0,0,...,0,0,0,0,0,0,1,0,1,0
2,1,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,1
3,0,0,0,1,0,0,0,0,0,0,...,0,0,0,0,0,0,0,0,0,0
4,0,0,0,0,0,0,0,0,0,0,...,0,0,0,1,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
3301,0,0,0,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,1,0
3302,0,0,0,0,0,0,0,0,0,0,...,0,1,0,0,0,0,0,0,1,0
3303,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,1,0,0,1
3304,0,0,0,0,0,0,0,0,0,0,...,0,0,0,0,0,0,1,1,0,0


In [130]:
longevity_data_catg_dum.columns

Index(['Country Name_Albania', 'Country Name_Algeria',
       'Country Name_American Samoa', 'Country Name_Andorra',
       'Country Name_Angola', 'Country Name_Antigua and Barbuda',
       'Country Name_Argentina', 'Country Name_Armenia',
       'Country Name_Australia', 'Country Name_Austria',
       ...
       'Country Code_ZWE', 'Region_East Asia & Pacific',
       'Region_Latin America & Caribbean', 'Region_Middle East & North Africa',
       'Region_North America', 'Region_South Asia',
       'Region_Sub-Saharan Africa', 'IncomeGroup_Low income',
       'IncomeGroup_Lower middle income', 'IncomeGroup_Upper middle income'],
      dtype='object', length=355)

In [131]:
longevity_data_catg_dum.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Columns: 355 entries, Country Name_Albania to IncomeGroup_Upper middle income
dtypes: int64(355)
memory usage: 9.0 MB


# Joining Split data of numerical data and categorical data converted to binary into one dataframe

In [132]:
longevity_prediction = pd.concat([longevity_scaled_data, longevity_data_catg_dum], axis=1)


In [133]:
longevity_prediction.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3306 entries, 0 to 3305
Columns: 365 entries, Year to IncomeGroup_Upper middle income
dtypes: float64(9), int64(356)
memory usage: 9.2 MB


#RUNNING MULTIPLE TREE BASED REGRESSION MODELS AND KERNEL REGRESSION MODEL

In [134]:
import xgboost as xgb
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

models = [
    LinearRegression(),
    KNeighborsRegressor(n_neighbors=1, weights='uniform'),
    KNeighborsRegressor(n_neighbors=10, weights='uniform'),
    KNeighborsRegressor(n_neighbors=100, weights='uniform'),
    DecisionTreeRegressor(max_depth=2),
    DecisionTreeRegressor(max_depth=5),
    DecisionTreeRegressor(max_depth=10),
    RandomForestRegressor(n_estimators=10, max_depth=5),
    RandomForestRegressor(n_estimators=100, max_depth=10),
    GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_depth=2),
    xgb.XGBRegressor(max_depth=2, learning_rate=0.1, n_estimators=100, min_child_weight=5),
]

# CROSS VALIDATION

CV is used for evaluating a model's performance and generalizability by repeatedly splitting the dataset into training and testing sets,
ensuring every data point gets used for both training and validation across iterations,  which provides a more reliable error estimate than a single split and helps detect overfitting


In [135]:
from sklearn.model_selection import cross_val_score, KFold

kf = KFold(n_splits=5, shuffle=True)

longevity_prediction.columns = longevity_prediction.columns.astype(str) # Convert column names to strings

for m in models:
    print(m)

    cv_scores = -cross_val_score(m, longevity_prediction, y, cv=kf, scoring='neg_mean_absolute_percentage_error')
    cv_scores = pd.DataFrame(cv_scores)
    display(cv_scores.describe().T)

    print()

LinearRegression()


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.013747,0.000884,0.012606,0.013432,0.013768,0.013878,0.015053



KNeighborsRegressor(n_neighbors=1)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.008586,0.000519,0.007775,0.008358,0.008855,0.008956,0.008985



KNeighborsRegressor(n_neighbors=10)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.020139,0.000822,0.01876,0.020164,0.02036,0.020463,0.020945



KNeighborsRegressor(n_neighbors=100)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.046314,0.001658,0.044201,0.045226,0.046492,0.04721,0.048439



DecisionTreeRegressor(max_depth=2)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.057478,0.002166,0.055182,0.056063,0.056611,0.059399,0.060134



DecisionTreeRegressor(max_depth=5)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.038619,0.001536,0.036864,0.037276,0.038845,0.039622,0.040489



DecisionTreeRegressor(max_depth=10)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.020365,0.001558,0.018138,0.019568,0.02077,0.021168,0.022182



RandomForestRegressor(max_depth=5, n_estimators=10)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.035335,0.001564,0.03296,0.0347,0.035909,0.03608,0.037025



RandomForestRegressor(max_depth=10)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.015733,0.000846,0.014439,0.015419,0.015942,0.016256,0.016611



GradientBoostingRegressor(max_depth=2)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.035123,0.00118,0.033951,0.034064,0.034955,0.036034,0.036612



XGBRegressor(base_score=None, booster=None, callbacks=None,
             colsample_bylevel=None, colsample_bynode=None,
             colsample_bytree=None, device=None, early_stopping_rounds=None,
             enable_categorical=False, eval_metric=None, feature_types=None,
             feature_weights=None, gamma=None, grow_policy=None,
             importance_type=None, interaction_constraints=None,
             learning_rate=0.1, max_bin=None, max_cat_threshold=None,
             max_cat_to_onehot=None, max_delta_step=None, max_depth=2,
             max_leaves=None, min_child_weight=5, missing=nan,
             monotone_constraints=None, multi_strategy=None, n_estimators=100,
             n_jobs=None, num_parallel_tree=None, ...)


,count,mean,std,min,25%,50%,75%,max
0,5.0,0.035296,0.001298,0.033654,0.034218,0.035844,0.036055,0.03671


# GRID SEARCH CROSS VALIDATION

Grid Search: Selects hyperparameters that minimize cross-validated error.

In [ ]:
from sklearn.model_selection import GridSearchCV

# Define models and their parameter grids
param_grids = {
    "KNeighborsRegressor": (
        KNeighborsRegressor(),
        {
            "n_neighbors": [1, 5, 10, 20, 50],
            "weights": ["uniform", "distance"]
        }
    ),

    "RandomForestRegressor": (
        RandomForestRegressor(),
        {
            "n_estimators": [10, 50, 100],
            "max_depth": [3, 5, 10],
            "min_samples_split": [2, 5, 10],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "GradientBoostingRegressor": (
        GradientBoostingRegressor(),
        {
            "n_estimators": [50, 100, 200],
            "learning_rate": [0.01, 0.1, 0.2],
            "max_depth": [2, 3, 5],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "DecisionTreeRegressor": (
        DecisionTreeRegressor(),
        {
            "max_depth": [2, 3, 5, 10],
            "min_samples_split": [2, 5, 10],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "LinearRegression": (
        LinearRegression(),
        {
            "fit_intercept": [True, False],
        }
    ),

    "XGBoost": (
        xgb.XGBRegressor(),
        {
            "n_estimators": [50, 100],
            "learning_rate": [0.01, 0.1, 0.2],
            "max_depth": [2, 3, 5],
            'min_child_weight': [1, 2, 5],
        }
    ),

  }

# Define cross-validation strategy
kf = KFold(n_splits=5, shuffle=True)

longevity_prediction.columns = longevity_prediction.columns.astype(str)

# Store results
results = []

# Perform GridSearchCV for each model
for name, (model, param_grid) in param_grids.items():
    print(f"Performing GridSearchCV for {name}...")

    grid_search = GridSearchCV(model, param_grid, cv=kf, scoring="neg_mean_absolute_percentage_error", n_jobs=-1)
    grid_search.fit(longevity_prediction, y)

    best_params = grid_search.best_params_
    best_score = grid_search.best_score_

    results.append({"Model": name, "Best Score": best_score, "Best Params": best_params})


results

Performing GridSearchCV for KNeighborsRegressor...
Performing GridSearchCV for RandomForestRegressor...


# PARSIMONY

Predict which features are important in the column using random forest.
Region has the highest feature in determining longevity.

In [ ]:
# Fit a Random Forest model
rf = RandomForestRegressor(n_estimators=1000)
rf.fit(longevity_prediction, y)

# Get feature importances
importances = pd.Series(rf.feature_importances_, index=longevity_prediction.columns).sort_values(ascending=False)
importances = importances[:20]

features_rf = importances.index

longevity_prediction_pars = longevity_prediction[ features_rf ]
importances

# Variance inflation factor (VIF) to detect multicollinearity among independent variables

VIF values that is 1.0  indicating no multicollinearity concern.
VIF values greater than 3.0 means multicollinearity concerns. Variables in column are dependent on one another (correlation)
VIF greeater than 5.0 will be removed.

In [ ]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
import statsmodels.api as sm

longevity_prediction_pars_with_const = sm.add_constant(longevity_prediction_pars)

vif = pd.Series(
    [variance_inflation_factor(longevity_prediction_pars_with_const.values, i) for i in range(longevity_prediction_pars_with_const.shape[1])],
    index=longevity_prediction_pars_with_const.columns
).sort_values(ascending=False)

vif[:20]

Dropping infinity VIF and VIF greater than 5

In [ ]:
longevity_prediction_vif = longevity_prediction_pars.copy()
while True:
  longevity_prediction_vif_with_const = sm.add_constant(longevity_prediction_vif)

  # Compute VIF for all features
  vif_series = pd.Series(
  [variance_inflation_factor(longevity_prediction_vif_with_const.values, i) for i in range(longevity_prediction_vif_with_const.shape[1])],
  index=longevity_prediction_vif_with_const.columns
    )
  #Exclude constant term
  vif_series = vif_series.drop('const', errors='ignore')
  highest_vif_feature = vif_series.idxmax()
  # Break the loop if all features have VIF ≤ 10
  if vif_series.loc[highest_vif_feature] <= 6:  # Checking the first feature after 'const'
    break

  # Drop the feature with the highest VIF
  longevity_prediction_vif = longevity_prediction_vif.drop(columns=[highest_vif_feature])

  print(f"Dropped: {highest_vif_feature} (VIF={vif_series.loc[highest_vif_feature]:.2f})")
  # end


vif_series[:25]

Running VIF after dropping multicollinear features

In [ ]:
from sklearn.model_selection import GridSearchCV

# Define models and their parameter grids
param_grids = {
    "KNeighborsRegressor": (
        KNeighborsRegressor(),
        {
            "n_neighbors": [1, 5, 10, 20, 50],
            "weights": ["uniform", "distance"]
        }
    ),

    "RandomForestRegressor": (
        RandomForestRegressor(),
        {
            "n_estimators": [10, 50, 100],
            "max_depth": [3, 5, 10],
            "min_samples_split": [2, 5, 10],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "GradientBoostingRegressor": (
        GradientBoostingRegressor(),
        {
            "n_estimators": [50, 100, 200],
            "learning_rate": [0.01, 0.1, 0.2],
            "max_depth": [2, 3, 5],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "DecisionTreeRegressor": (
        DecisionTreeRegressor(),
        {
            "max_depth": [2, 3, 5, 10],
            "min_samples_split": [2, 5, 10],
            'min_samples_leaf': [1, 2, 5, 10]
        }
    ),

    "LinearRegression": (
        LinearRegression(),
        {
            "fit_intercept": [True, False],
        }
    ),

      "XGBoost": (
        xgb.XGBRegressor(),
        {
            "n_estimators": [50, 100],
            "learning_rate": [0.01, 0.1, 0.2],
            "max_depth": [2, 3, 5],
            'min_child_weight': [1, 2, 5],
        }
    ),

  }

# Define cross-validation strategy
kf = KFold(n_splits=5, shuffle=True)

longevity_prediction.columns = longevity_prediction.columns.astype(str) # Convert column names to strings

# Store results
results = []

# Perform GridSearchCV for each model
for name, (model, param_grid) in param_grids.items():
    print(f"Performing GridSearchCV for {name}...")

    grid_search = GridSearchCV(model, param_grid, cv=kf, scoring="neg_mean_absolute_percentage_error", n_jobs=-1)
    grid_search.fit(longevity_prediction_vif, y)

    best_params = grid_search.best_params_
    best_score = grid_search.best_score_

    results.append({"Model": name, "Best Score": best_score, "Best Params": best_params})
# end

results

*****************************************************************************

#RUNNING DEEP LEARNING: KERAS MODEL TO FIND TRAINING AND TEST VALIDATION FOR LONGEVITY

*****************************************************************************

# Performing KERAS CROSS VALIDATION.

In [ ]:
from keras.models import Sequential
from keras.layers import *
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.optimizers.schedules import ExponentialDecay, PolynomialDecay
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

def build_model( longevity_prediction_train, epochs, batch_size, decay_frac, initial_learning_rate, dropout_rate ):

    # Create model
    model = Sequential([
        Input(shape=longevity_prediction_train.shape[1]),
        Dense(128, activation='relu'),
        Dropout(dropout_rate),
        Dense(64, activation='relu'),
        Dropout(dropout_rate),
        Dense(256, activation='relu'),
        Dense(1, activation='relu'),
    ])

    # define training parameters
    decay_epochs    = int( epochs * decay_frac )
    n_train         = longevity_prediction_train.shape[0]
    steps_per_epoch = int( np.ceil( n_train / batch_size ) )
    decay_steps     = decay_epochs * steps_per_epoch

    lr_schedule_poly = PolynomialDecay(
        initial_learning_rate=initial_learning_rate,
        end_learning_rate=0.001,
        decay_steps=decay_steps,
        power=2.0
    )

    # Define the optimizer with a custom learning rate
    optimizer = Adam(
        learning_rate=lr_schedule_poly,
    )

    # Compile model
    model.compile(
        optimizer=optimizer,
        loss='mae'
    )

    return model

**Mean Absolute Error vs negative Mean Absolute Error**  Optimization algorithms (like those used for cross-validation and hyperparameter tuning ) are designed to maximize a score, whereas Mean Absolute Error (MAE) is an "error" metric that needs to be minimized.

*   MAE: How far off your model's predictions typically are from the actual values.
*   Interpretation: A lower MAE is better, indicating smaller average prediction errors




*   𝑅2 : How much of the variability in your data your model explains.
*   Interpretation: Higher values (closer to 1) are better, meaning the model explains more variance. An R^2 𝑅2 of 0.7 means the model explains 70% of the data's variation.




ML Scoring Convention: Machine learning libraries follow a general convention where "higher return values are better" for all scoring metrics.

To use MAE (an error metric where lower values are better) within this "maximization" framework, the value is negated.
This creates the NMAE (Negative Mean Absolute Error)


In [ ]:
from sklearn.model_selection import KFold
from sklearn.metrics import mean_absolute_error, r2_score

# Re-defining build_model with the fix for the Input layer
from keras.models import Sequential
from keras.layers import *
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.optimizers.schedules import ExponentialDecay, PolynomialDecay
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

def build_model( longevity_prediction_train, epochs, batch_size, decay_frac, initial_learning_rate, dropout_rate ):

    # Create model
    model = Sequential([
        Input(shape=(longevity_prediction_train.shape[1],)),
        Dense(128, activation='relu'),
        Dropout(dropout_rate),
        Dense(64, activation='relu'),
        Dropout(dropout_rate),
        Dense(256, activation='relu'),
        Dense(1, activation='relu'),
    ])

    # define training parameters
    decay_epochs    = int( epochs * decay_frac )
    n_train         = longevity_prediction_train.shape[0]
    steps_per_epoch = int( np.ceil( n_train / batch_size ) )
    decay_steps     = decay_epochs * steps_per_epoch

    lr_schedule_poly = PolynomialDecay(
        initial_learning_rate=initial_learning_rate,
        end_learning_rate=0.001,
        decay_steps=decay_steps,
        power=2.0
    )

    # Define the optimizer with a custom learning rate
    optimizer = Adam(
        learning_rate=lr_schedule_poly,
    )

    # Compile model
    model.compile(
        optimizer=optimizer,
        loss='mae'
    )

    return model
# end

epochs                = 500
batch_size            = 256
decay_frac            = 0.8
dropout_rate          = 0.1
initial_learning_rate = 0.01

n_splits = 5
kf = KFold(n_splits=n_splits, shuffle=True)

train_mae_list = []
test_mae_list = []
train_r2_list = []
test_r2_list = []

fold = 1
for train_index, test_index in kf.split(longevity_prediction):
    print(f"Fold {fold}/{n_splits}")
    longevity_prediction_train, longevity_prediction_test = longevity_prediction.iloc[train_index], longevity_prediction.iloc[test_index]
    y_train, y_test = y.iloc[train_index], y.iloc[test_index]

    # Build a fresh model (which resets optimizer state including the learning rate schedule)
    model = build_model(
        longevity_prediction_train,
        epochs,
        batch_size,
        decay_frac,
        initial_learning_rate,
        dropout_rate
    )

    # Early stopping callback
    early_stopping = EarlyStopping(
        monitor='val_loss',  # Monitor validation loss
        patience=100,          # Stop after 5 epochs without improvement
        restore_best_weights=True  # Restore the best weights after stopping
    )

    # Train the model both x and y
    history = model.fit(
        longevity_prediction_train, y_train,
        epochs=epochs,
        batch_size=batch_size,
        validation_data=(longevity_prediction_test, y_test),
        verbose=0,
        callbacks=[early_stopping]
    )

    # Evaluate on training data
    y_train_pred = model.predict(longevity_prediction_train, verbose=0)
    train_mae = mean_absolute_error(y_train, y_train_pred)
    train_r2 = r2_score(y_train, y_train_pred)

    # Evaluate on test data
    y_test_pred = model.predict(longevity_prediction_test, verbose=0)
    test_mae = mean_absolute_error(y_test, y_test_pred)
    test_r2 = r2_score(y_test, y_test_pred)

    print("MAE:")
    print(f"  Train: {train_mae:.4f}")
    print(f"  Test:  {test_mae:.4f}")
    print("R^2:")
    print(f"  Train: {train_r2:.4f}")
    print(f"  Test:  {test_r2:.4f}")
    print()

    train_mae_list.append(train_mae)
    test_mae_list.append(test_mae)
    train_r2_list.append(train_r2)
    test_r2_list.append(test_r2)

    fold += 1
# end

In [ ]:
cross_validation_result_df = pd.DataFrame({
    'Train MAE': train_mae_list,
    'Test MAE': test_mae_list,
    'Train R2': train_r2_list,
    'Test R2': test_r2_list
})

cross_validation_result_df.describe()

# Train test split 80% : 20%

In [ ]:
from sklearn.model_selection import train_test_split

longevity_prediction_train, longevity_prediciton_test, y_train, y_test = train_test_split(longevity_prediction_vif, y,
                                                    test_size=0.2)

# Ensure train/test label distributions are as similar as possible (stratification for regression)

In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt

# Define the number of attempts
attempts = 1000

# Determine the global min and max of y
y_min, y_max = y.min(), y.max()
bins = np.linspace(y_min, y_max, 25)  # 25 bins means 26 edges

best_mse = float('inf')
best_split = None

for i in range(attempts):
    # Perform train-test split
    longevity_prediction_train_i, longevity_prediction_test_i, y_train_i, y_test_i = train_test_split(longevity_prediction_vif, y, test_size=0.2)

    # Compute histograms
    y_train_hist, _ = np.histogram(y_train_i, bins=bins, density=True)
    y_test_hist, _ = np.histogram(y_test_i, bins=bins, density=True)

    # Compute MSE between histograms
    mse = mean_squared_error(y_train_hist, y_test_hist)

    # Save the best split
    if mse < best_mse:
        best_mse = mse
        best_split = (longevity_prediction_train_i, longevity_prediction_test_i, y_train_i, y_test_i)

# Unpack the best split
longevity_prediction_train, longevity_prediction_test, y_train, y_test = best_split

# Display histograms of the best split
plt.hist(y_train, bins=bins, alpha=0.5, color='b', label='Train', density=True)
plt.hist(y_test, bins=bins, alpha=0.3, color='r', label='Test', density=True)
plt.legend()
plt.xlabel('y values')
plt.ylabel('Density')
plt.title('Best Train-Test Split Histogram Comparison')
plt.show()

# Creating Deep Learning KERAS Model

In [ ]:
from keras.models import Sequential
from keras.layers import *
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.callbacks import EarlyStopping

dropout_rate = 0.05 #I changed dropout rate from 0.05 to 0.01 used to prevent overfitting

# Create model complexity
model = Sequential([
    Input(shape=(longevity_prediction_train.shape[1],)),
    Dense(64, activation='relu'), # originally 32 neurons
    BatchNormalization(),
    Dropout(dropout_rate),
    Dense(128, activation='relu'), # originally 64 neurons
    BatchNormalization(),
    Dropout(dropout_rate),
    Dense(32, activation='relu'), #originally 128 neurons
    Dense(1, activation='relu'),
])

# Define the optimizer with a custom learning rate
optimizer = Adam(learning_rate=0.01)

# Compile model
model.compile(
    optimizer=optimizer,
    loss='mae'
)

# Early stopping callback
early_stopping = EarlyStopping(
    monitor='val_loss',  # Monitor validation loss
    patience=100,          # Stop after 5 epochs without improvement, changed from 100 to 50
    restore_best_weights=True  # Restore the best weights after stopping
)

# Display model summary
model.summary()

# Create learning rate schedule

In [ ]:
from tensorflow.keras.optimizers import Adam
from tensorflow.keras.optimizers.schedules import ExponentialDecay, PolynomialDecay

# define training parameters
epochs          = 500 # Originally 500
batch_size      = 256
decay_epochs    = int( epochs * 0.5 )

n_train         = longevity_prediction_train.shape[0]
steps_per_epoch = int( np.ceil( n_train / batch_size ) )
decay_steps     = decay_epochs * steps_per_epoch
#print(decay_steps)

# Define the learning rate schedule
initial_learning_rate = 0.05

lr_schedule_exp = ExponentialDecay(
    initial_learning_rate=initial_learning_rate,
    decay_rate=5.0,
    decay_steps=decay_steps
)

lr_schedule_poly = PolynomialDecay(
    initial_learning_rate=initial_learning_rate,
    end_learning_rate=0.001,
    decay_steps=decay_steps,
    power=1.0
)

# Define the optimizer with a custom learning rate
optimizer = Adam(
    learning_rate=initial_learning_rate,
    #learning_rate=lr_schedule_exp,
    #learning_rate=lr_schedule_poly,
    #clipnorm=1,
    #clipvalue=1,
)

# Compile model
model.compile(
    optimizer=optimizer,
    loss='mae'
)

# Adding early stopping and learning rate reduction callbacks

In [ ]:
from tensorflow.keras.callbacks import EarlyStopping, ReduceLROnPlateau

# Early stopping callback
early_stopping = EarlyStopping(
    monitor='val_loss',  # Monitor validation loss
    patience=30,          # Stop after 50 epochs without improvement
    restore_best_weights=True  # Restore the best weights after stopping
)

# Callback to reduce LR when a monitored metric has stopped improving
reduce_lr = ReduceLROnPlateau(
    monitor='val_loss',       # metric to monitor
    factor=0.5,               # factor by which to reduce the LR
    patience=30,              # number of epochs with no improvement after which LR will be reduced, changed from 50 to 40
    min_lr=1e-6,              # lower bound on the learning rate
    verbose=1                 # print when LR is reduced
)

# Training KERAS model

In [ ]:
# Train model
history = model.fit(
    longevity_prediction_train, y_train,
    epochs=400, #originally at 400
    batch_size=batch_size,
    shuffle=True,
    validation_data=(longevity_prediciton_test, y_test),
    verbose=1,
    callbacks=[
        early_stopping,
    ]
)

In [ ]:
from sklearn.metrics import mean_absolute_error, r2_score

# Get predictions
y_train_pred = model.predict(longevity_prediction_train)[:,0]
y_test_pred  = model.predict(longevity_prediction_test)[:,0]

# Create a figure with two subplots side by side
fig, axes = plt.subplots(1, 2, figsize=(20, 8))

# Left subplot: Training and Validation Loss
axes[0].plot(history.history['loss'], label='Training Loss')
axes[0].plot(history.history['val_loss'], label='Validation Loss')
axes[0].set_yscale('log')
axes[0].set_title('Training and Validation Loss')
axes[0].set_xlabel('Epochs')
axes[0].set_ylabel('Loss')
axes[0].legend()

# Right subplot: True vs Predicted Data
axes[1].scatter(y_train, y_train_pred, c='b', s=1, label="Train set")
axes[1].scatter(y_test, y_test_pred, c='r', s=1, label="Test set")
axes[1].set_title('True vs Pred (test data)')
axes[1].set_xlabel('true')
axes[1].set_ylabel('pred')
axes[1].legend()

plt.tight_layout()
plt.show()

mae_train = mean_absolute_error(y_train, y_train_pred)
r2_train = r2_score(y_train, y_train_pred)
mae_test = mean_absolute_error(y_test, y_test_pred)
r2_test = r2_score(y_test, y_test_pred)

print("MAE:")
print(f"  Train: {mae_train:.4f}")
print(f"  Test:  {mae_test:.4f}")
print("R^2:")
print(f"  Train: {r2_train:.4f}")
print(f"  Test:  {r2_test:.4f}")

# Comparison with other model (KERAS versus TREE MODEL)

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor
from sklearn.tree import DecisionTreeRegressor
import xgboost as xgb
from sklearn.metrics import mean_absolute_error, r2_score

# Define regression models corresponding to the original classifiers
models = [
    LinearRegression(),
    KNeighborsRegressor(n_neighbors=5, weights='uniform'),
    KNeighborsRegressor(n_neighbors=10, weights='uniform'),
    KNeighborsRegressor(n_neighbors=50, weights='uniform'),
    RandomForestRegressor(n_estimators=100, max_leaf_nodes=3),
    RandomForestRegressor(n_estimators=100, max_leaf_nodes=10),
    RandomForestRegressor(n_estimators=100, max_leaf_nodes=30),
    GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_leaf_nodes=3),
    GradientBoostingRegressor(n_estimators=100, learning_rate=0.1, max_leaf_nodes=10),
    DecisionTreeRegressor(max_leaf_nodes=3),
    DecisionTreeRegressor(max_leaf_nodes=10),
    xgb.XGBRegressor(max_depth=2, learning_rate=0.1, n_estimators=100, min_child_weight=5),
]

In [ ]:
for m in models:
    m.fit(longevity_prediction_train, y_train)
    y_train_pred = m.predict(longevity_prediction_train)
    y_test_pred = m.predict(longevity_prediction_test)

    # Compute regression metrics
    mae_train = mean_absolute_error(y_train, y_train_pred)
    r2_train = r2_score(y_train, y_train_pred)
    mae_test = mean_absolute_error(y_test, y_test_pred)
    r2_test = r2_score(y_test, y_test_pred)

    # Print model and its metrics
    print(m)
    print("MAE:")
    print(f"  Train: {mae_train:.4f}")
    print(f"  Test:  {mae_test:.4f}")
    print("R^2:")
    print(f"  Train: {r2_train:.4f}")
    print(f"  Test:  {r2_test:.4f}")
    print()
# end

# Interpretation of the true vs pred data from graph to table to give a numerical data of longevity

In [ ]:
# Create a DataFrame to organize the data of true vs pred data graph of a R of 0.83 which is a bit low
predicted_results_df = pd.DataFrame({
    'Actual Life Expectancy': y_test,
    'Predicted Life Expectancy': y_test_pred,
    'Difference (Error)': (y_test - y_test_pred),
    'Percentage Error (%)': ((abs(y_test - y_test_pred) / y_test) * 100)
})

# Get the original 'Country Name' and 'Year' directly from data_drop using y_test's index
country_year_info = data_drop.loc[y_test.index, ['Country Name', 'Year']]

# Merge this information into predicted_results_df, aligning by index
predicted_results_df = predicted_results_df.merge(country_year_info, left_index=True, right_index=True)

# Reorder columns to have Country Name and Year at the beginning for clarity
predicted_results_df = predicted_results_df[[
    'Country Name', 'Year', 'Actual Life Expectancy', 'Predicted Life Expectancy',
    'Difference (Error)', 'Percentage Error (%)'
]]

# Display the resulting numerical table without explicit sorting, showing the split order
predicted_results_df

In [ ]:
predicted_results_df.info()

There is 662 entries which represent 0.2 or 20% of the test data done during the 80% : 20% data split from the original y test variable of the test data (life expectancy) which had 3306 rows. 3306 * 0.2 gives 661.2 rows

In [ ]:
# Filter predicted_results_df for 'Countries' to see all its entries in the test set
countries_predictions_corrected = predicted_results_df[predicted_results_df['Country Name'] == 'Germany'].sort_values(by='Year')
print("Country entries in the corrected predicted_results_df (from original csv dataset):")
display(countries_predictions_corrected)

# Display the specific entry for Countries in 2014 from the original data_drop
print("\nLife Expectancy for random country in 2014 from original data_drop:")
display(data_drop[(data_drop['Country Name'] == 'Germany') & (data_drop['Year'] == 2016)])

# Display all data_drop entries with Life Expectancy around 45.100
print("\nAll data_drop entries with Life Expectancy around 45.100:")
display(data_drop[data_drop['Life Expectancy World Bank'].between(45.0, 45.2)])

In [ ]:
# Get the original indices from y_test (which was used to create predicted_results_df)
original_indices = y_test.index

# Map these indices back to data_drop to get the 'Country Name'
country_names_for_test_set = data_drop.loc[original_indices, 'Country Name']

# Add 'Country Name' to the predicted_results_df
predicted_results_df['Country Name'] = country_names_for_test_set.values

predicted_results_df = predicted_results_df[['Country Name', 'Year','Actual Life Expectancy', 'Predicted Life Expectancy', 'Difference (Error)', 'Percentage Error (%)']]

# Display the updated DataFrame
display(predicted_results_df.head())

What will likely most explain the prediction trend seen in the predicted life expectaancy? It tends to be random some times it goes up and sometimes it gorees down

# SHAP (Output for ML)

SHAP TREE BASED MODEL

In [ ]:
import shap

In [ ]:
from sklearn.linear_model import LinearRegression
model = LinearRegression().fit(longevity_prediction_train, y_train)

# 2. Incorporate SHAP (No re-training needed)
import shap
explainer = shap.Explainer(m, longevity_prediction_train) # It just looks at your existing 'model'
shap_values = explainer(longevity_prediction_test)

# 3. Visualize
shap.plots.waterfall(shap_values[0])

Waterfall Plot (shap.plots.waterfall): This plot is used for explaining individual predictions. It starts from the model's base value (the average prediction) and shows how each feature's SHAP value pushes the prediction up (red) or down (blue) to reach the final output value for a single instance.

Base Value (E[f(x)]): This is the average output of the model over the entire dataset. It's the starting point for the explanation.
Features (e.g., Region_Sub-Saharan Africa, Prevalence of Undernourishment): Each bar represents a feature from your input data.

Color of the Bars:
Red bars (positive SHAP values) indicate features that pushed the prediction higher than the base value.
Blue bars (negative SHAP values) indicate features that pushed the prediction lower than the base value.
Length of the Bars: The length of each bar shows the magnitude of the feature's impact on the prediction.
Output Value (f(x)): This is the final prediction of the model for the specific instance being explained.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
model = RandomForestRegressor().fit(longevity_prediction_train, y_train)

# 2. Incorporate SHAP (No re-training needed)
import shap
explainer = shap.Explainer(m, longevity_prediction_train) # It just looks at your existing 'model'
shap_values = explainer(longevity_prediction_test)

# 3. Visualize
shap.plots.waterfall(shap_values[0])

Summary Plot (shap.plots.summary): This plot provides a global view of feature importance and impact. It shows, for each feature:

Its overall importance (features are typically ordered by their mean absolute SHAP value).
The distribution of SHAP values for that feature across the dataset.
How the feature's value (color: low to high) influences the prediction (positive or negative SHAP values).

In [ ]:
import shap

explainer = shap.TreeExplainer(model)

shap_values = explainer.shap_values(longevity_prediction_test)

shap.summary_plot(shap_values, longevity_prediction_test)

In [ ]:
import shap
explainer = shap.TreeExplainer(model)

shap_values = explainer.shap_values(longevity_prediction_test)

shap.decision_plot(explainer.expected_value, shap_values, longevity_prediction_test)

SHAP KERNEL BASED MODEL

In [ ]:

from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor

Decision Plot (shap.decision_plot): This plot shows how an individual prediction is reached, much like a waterfall plot, but it arranges features in a different way, often showing the cumulative sum of SHAP values from a baseline to the final prediction

*******************************************************************************

# VISUALIZATION PORTION

******************************************************************************

Visualization of complete dataset with nan filled using data_drop

Data inspection re-visited

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

%matplotlib inline

In [ ]:
catg_data_drop = data_drop.select_dtypes(include=object)
catg_data_drop.info()

In [ ]:
num_data_drop = data_drop.select_dtypes(exclude=object)
num_data_drop.info()

# Comparison Plot



*   Line Charts
*   Bar Charts
*   Radar Charts



In [ ]:
life_expectancy_mean_global = data_drop.groupby('Year')['Life Expectancy World Bank'].mean()
life_expectancy_mean_global

# Plotting panda Series vs Dataframe

life_expectancy_mean_global is a Pandas Series, not a DataFrame. A Series fundamentally consists of two parts:

An Index: This acts as the label for each data point. In our life_expectancy_mean_global Series, the 'Year' is stored as its index.
Values: These are the actual data points.

By explicitly passing life_expectancy_mean_global.index for the x-axis and life_expectancy_mean_global.values for the y-axis, we ensure that matplotlib correctly understands which part of the Series should be plotted on which axis.

If we tried to treat it like a DataFrame column named 'Year' (e.g., plt.plot('Year', data=life_expectancy_mean_global)), it wouldn't work because the Series itself doesn't have a 'Year' column; 'Year' is its index.

In [ ]:
# Target variable (Life Expectancy World Bank) trend across year for all region

plt.figure(figsize=(16, 8), dpi=300)

# Plot data
plt.plot(life_expectancy_mean_global.index, life_expectancy_mean_global.values, marker='o', linestyle='-', color='b')

# Specify ticks for x- and y-axis
plt.xticks(np.arange(2000, 2020, 1), rotation=70)
plt.yticks(np.arange(65, 75, 1))

# Add title and label for y-axis
plt.title('Plot of mean global life Expectancy with respect to year from 2001 to 2019 ', fontsize=16)
plt.ylabel('Life Expectancy', fontsize=14)

# Add grid
plt.grid()

# Add legend
plt.legend(loc='upper left')

# Show plot
plt.show()

Positive trend increase seen in the global life expectancy betwen 2001 to 2019.

In [ ]:
# Taking the mean of life expectancy for each year across all region
life_expectancy_mean_region_global = data_drop.groupby('Year')['Life Expectancy World Bank'].mean()
life_expectancy_mean_region_global

In [ ]:
# Confirmation of life expectancy globally from 2001 to 2019
np.mean(data_drop['Life Expectancy World Bank'])

In [ ]:
# Histogram Plot of mean life expectancy from 2001 - 2019

In [ ]:
# Histogram Plot of mean life expectancy from 2001 - 2019
life_expectancy_mean_region_global = data_drop.groupby('Region')['Life Expectancy World Bank'].mean()
life_expectancy_mean_region_global

In [ ]:
# Looking at communicable and noncommunicable diseases per year on all region
life_expectancy_mean_region_global_disease = data_drop.iloc[:, [2, 4, 5, 13, 14]].groupby(['Region', 'Year', 'NonCommunicable'])['Life Expectancy World Bank'].mean()
life_expectancy_mean_region_global_disease

In [ ]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(life_expectancy_mean_region_global_disease)

NOTE: I can change the region and year of the code below and look at the mean of all countries in that region per year and then compare it to the mean of all regions. For visualization use a region that shows great amount of disparity among their respective neighboring countries and then add some more column such as noncommunicable disease and health care expenditure, communicable disease and sanitation or unemployment and undernourishment

In [ ]:
region_year = data_drop.iloc[:, [0, 2, 4, 5]]
East_Asia_year = region_year[(region_year['Region'] == 'East Asia & Pacific') & (region_year['Year'] == 2002)]
East_Asia_year

In [ ]:
np.mean(East_Asia_year['Life Expectancy World Bank'])

Inspecting role of communicable disease and longevity

In [ ]:
communicable_year = data_drop.iloc[:, [0, 2, 4, 5, 11, 13]]
Latin_America_year = communicable_year[(communicable_year['Region'] == 'Latin America & Caribbean') & (communicable_year['Year'] == 2008)].sort_values('Communicable')
Latin_America_year

In [ ]:
np.mean(Latin_America_year['Life Expectancy World Bank'])

Inspecting unemployment and undernourishment

In [ ]:
nourishment_year = data_drop.iloc[:, [0, 2, 4, 5, 6, 9, 10]]
Sub_Saharan_year = nourishment_year[(nourishment_year['Region'] == 'Sub-Saharan Africa') & (nourishment_year['Year'] == 2001)].sort_values('Unemployment')
Sub_Saharan_year

In [ ]:
np.mean(Sub_Saharan_year['Life Expectancy World Bank'])

In [ ]:
print('Number of country in Sub Saharan Africa is:' , nourishment_year[nourishment_year['Region'] == 'Sub-Saharan Africa']['Country Name'].nunique())

NonCommunicable diseases such as cancer, health care expenditure and longevity

In [ ]:
noncommunicable_year = data_drop.iloc[:, [0, 2, 4, 5, 7, 8, 14]]
noncommunicable_year = noncommunicable_year[(noncommunicable_year['Region'] == 'South Asia') & (noncommunicable_year['Year'] == 2001)].sort_values('Health Expenditure %')
noncommunicable_year

boolean indexing or boolean masking.
This expression doesn't directly give you data; it creates a Series of True and False values. Each True corresponds to a row where the 'CO2' value is greater than 90000, and each False corresponds to a row where it's not.

The outer part: data_drop[...]

When you put this boolean Series inside the square brackets of the data_drop DataFrame (e.g., data_drop[your_boolean_series]), pandas uses that mask to select rows.


In [ ]:

CO2 = data_drop.iloc[:, [0, 2, 4, 5, 7, 8, 14]]
CO2_long = CO2[(data_drop['CO2'] > 300000) & (CO2['Year'] == 2001)].sort_values(['Life Expectancy World Bank'])
CO2_long

In [ ]:
np.mean(data_drop['CO2'])

In [ ]:
x = data_drop['Region'].unique()
x

In [ ]:
#Using this to determine the column (irrelevant and redudant)
data_drop.info()

In [ ]:
# NEED TO UPDATE HISTOGRAM BECAUSE HISTOGRAM IS BETWEEN TWO NUMERICAL VARIABLE, I HAVE AN OBJECT CALLED REGION

# Create figure
plt.figure(figsize=(6, 4), dpi=150)

# Create histogram
plt.hist(life_expectancy_mean_region_global, bins=7)
plt.axvline(x=65, color='g')
plt.axvline(x=60, color='g', linestyle= '--')
plt.axvline(x=75, color='g', linestyle= '--')

# Add labels and title
plt.xlabel('Life Expectancy Mean')
plt.ylabel('Region')
plt.title('Global Mean Life Expectancy by Region from 2001 to 2019')

# Add grid
plt.grid(True)

# Show plot
plt.show()


#**Using groupby for a DATAFRAME**

In [ ]:
life_expectancy_mean_region_global_per_year =  data_drop.iloc[:, [2, 4, 5]].groupby(['Region', 'Year'])['Life Expectancy World Bank'].mean()
life_expectancy_mean_region_global_per_year

In [ ]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(life_expectancy_mean_region_global_per_year)

In [ ]:
# Target variable (Life Expectancy World Bank) trend across year for all region

# Make a random dataset:
height = (life_expectancy_mean_region_global.values)
bars = (life_expectancy_mean_region_global.index)
y_pos = np.arange(len(bars))

# Create bars
plt.bar(y_pos, height, color = "green")

# Create names on the x-axis
plt.xticks(y_pos, bars, rotation=90)

plt.title('Mean Life Expectancy trend across year for all region from 2001 to 2019')
plt.xlabel('Average Life Expectancy Across all year')
plt.ylabel('Region')
plt.legend(loc = 'upper left')

# Show graphic
plt.show()

# Relation Plot



*   Scatter Plot
*   Heatmaps
*   Correlogram
*   Bubble Plots





# Composition Plot

In [ ]:
catg_data_drop.info()

In [ ]:
region_pie = data_drop.groupby(['Region', 'Country Name'])['Country Code'].count()
region_pie

In [ ]:
region_income = catg_data_drop[['Region', 'IncomeGroup']].groupby(['Region'])
count_region = region_income.value_counts()
count_region

In [ ]:
# Create figure
plt.figure(figsize=(16, 9), dpi=300)

# Create pie plot
plt.pie(region_pie, labels=count_region.index, data=count_region.values, autopct='%.0f%%', pctdistance = 0.85, startangle = 85, rotatelabels=True)

# Add title
label = 'Percentage of Income Class across Region'
#plt.title('Percentage of Income Class across Region')

# add a white circle at the center
my_circle=plt.Circle( (0,2), 0.7)
p=plt.gcf()
p.gca().add_artist(my_circle)
plt.show()
# Show plot
plt.show()

# Using groupby for a series

In [ ]:
# Create figure
plt.figure(figsize=(16, 9), dpi=300)

# Create pie plot
plt.pie(count_region, labels=count_region.index, data=count_region.values, autopct='%.0f%%', pctdistance = 0.85, startangle = 85, rotatelabels=True)

# Add title
label = 'Percentage of Income Class across Region'
#plt.title('Percentage of Income Class across Region')

# add a white circle at the center
my_circle=plt.Circle( (0,2), 0.7)
p=plt.gcf()
p.gca().add_artist(my_circle)
plt.show()
# Show plot
plt.show()


In [ ]:
region_income = catg_data_drop[['Country Name', 'Region', 'IncomeGroup']].groupby('IncomeGroup')
count_region = region_income.value_counts()
count_region

In [ ]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(count_region)

In [ ]:
num_data_drop.info()

In [ ]:
health_life_expect = data_drop[['Health Expenditure %', 'Life Expectancy World Bank', 'Year', 'Region', 'Country Name']].groupby(['Year', 'Region'])
health_life_expect = health_life_expect.value_counts()
health_life_expect

In [ ]:
idx = pd.IndexSlice
health_trend = health_life_expect.loc[idx[:, 'North America', :, :]]
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(health_trend)

In [ ]:
education_life_expect = data_drop[['Education Expenditure %', 'Life Expectancy World Bank', 'Year', 'Region', 'Country Name']].groupby(['Year', 'Region'])
education_life_expect = education_life_expect.value_counts()
education_life_expect

In [ ]:
idx = pd.IndexSlice
education_trend = education_life_expect.loc[idx[:, 'North America', :, :]]
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(education_trend)

In [ ]:
undernourish_life_expect = data_drop[['Prevelance of Undernourishment', 'Life Expectancy World Bank', 'Year', 'Region', 'Country Name']].groupby(['Year', 'Region'])
undernourish_life_expect = undernourish_life_expect.value_counts()
undernourish_life_expect

In [ ]:
idx = pd.IndexSlice
region_trend = undernourish_life_expect.loc[idx[:, 'South Asia', :, :]]
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    display(region_trend)



*   Pie Charts
*   Stacked Area Charts
*   Venn Diagrams









A Pandas Series inherently has only one 'column' of values, along with its index. It's a single, one-dimensional data structure.
You use unstack() when your Series has a MultiIndex, meaning its index itself has multiple levels (like 'Region' and 'Year' in our life_expectancy_mean_region_year example).

In [ ]:
life_expectancy_mean_region_year = data_drop.groupby(['Region', 'Year'])['Life Expectancy World Bank'].mean()
life_expectancy_mean_region_year

In [ ]:
# Create figure
plt.figure(figsize=(12, 8), dpi=300)


life_expectancy_by_region_year = life_expectancy_mean_region_year.unstack(level='Region')

# Prepare x and y values for stackplot
x_values = life_expectancy_by_region_year.index
y_values = [life_expectancy_by_region_year[col].values for col in life_expectancy_by_region_year.columns]
labels = life_expectancy_by_region_year.columns

# Create stacked area chart
plt.stackplot(x_values, *y_values, labels=labels)

# Add legend
plt.legend(title='Region', bbox_to_anchor=(1.05, 1), loc='upper left')

# Add labels and title
plt.xlabel('Year')
plt.ylabel('Life Expectancy World Bank')
plt.title('Mean Life Expectancy Trend by Region (2001-2019)')

# Add ticks
plt.xticks(np.arange(2000, 2020, 1), rotation=70)
plt.yticks(np.arange(65, 75, 1))

# Add grid
plt.grid(True)

# Show plot
plt.tight_layout()
plt.show()

# Distribution Plot



*   Histogram
*   Density Plot
*   Box Plot
*   Violin Plot





# Geo Plots



*   Connection Maps
*   Cartograms
*   Dot Maps




